src/vs/vscode.proposed.d.ts

```diff
@@ -1181,6 +1181,45 @@ declare module 'vscode' {
 	}
 	//#endregion
 
+	/**
+	 * Class used to execute an extension callback as a task.
+	 */
+	export class CustomExecution {
+		/**
+		 * @param callback The callback that will be called when the extension callback task is executed.
+		 */
+		constructor(callback: (terminalRenderer: TerminalRenderer, cancellationToken: CancellationToken, thisArg?: any) => Thenable<number | undefined>);
```

Comment:
I think this should be `Thenable<number>` because it's acting as a process, the only way for undefined to fire in `onExit` is by disposing the terminal which a `CustomExecution` cannot do. @alexr00?

Reply:
That sounds good to me.

Reply:
I am not sure I agree. You cannot assume that a custom execution is running a process at all.

The architecture in VSCode wants to "tie" a terminal (console) to a process, but even that isn't what a terminal is supposed to be. A terminal is simply a mechanism for receiving input and displaying output. What that terminal is displaying input\output for should be an abstract concept (just like you had many VT100 terminals connected to a VAX system - showing my age).

When we are finished with this PR, "ShellExecution" and "ProcessExecution" could be implemented as a "CustomExecution" (and I'm sure folks will do this for custom shells) and is it stands CustomExecution doesn't tie itself to a process.

But, for the sake of moving forward with this PR, if you want a CustomExecution to complete with a number, then that's okay, but I suspect folks will return values like "42" since it doesn't make any sense in some circumstances.

Reply:
@Tyriar and @alexr00 - I'd like your thoughts on this.

Reply:
> but I suspect folks will return values like "42" since it doesn't make any sense in some circumstances.

But the documentation would read "use 0 for success and non-zero for failure", 42 would be fine provided the task failed, no?

Reply:
> The architecture in VSCode wants to "tie" a terminal (console) to a process, but even that isn't what a terminal is supposed to be.

xterm.js provides a terminal, vscode stitches it together with a process. I don't see why we wouldn't want to continue this as it would simplify things and minimize changes needed in the terminal part.

Reply:
@Tyriar  It's an architectural decision for VSCode engineers to make. I easily envision terminals being used for purposes other than communicating with a process which is really what this change allows since a custom execution could in theory run forever and never actually communicate with any process.

That being said, I'm okay with going back to just a number and updating the documentation.

Reply:
Let's keep the two concepts tied, and have custom execution pretend to be a process. I think that's the simplest approach, we can easily change it later. We could even move to `Thenable<number | undefined>` in the API later if it made sense, we couldn't go the other way though as it's a breaking change.

Reply:
@Tyriar - This has been addressed